# Mishra Data Quality Index 1 (1-2-3.ipynb)
Zois Boukouvalas, Max Calzada 
December 1, 2022 (Latest Update)

## What does DQI c1 do?

DQI c1 measures the following:

*   unique words per total number of data samples in the dataset
*   relative amount of sentences within appropriate legnth (3-30 words)*
*   stantard deviation of sentence lengths

DQI c1 assigns a higher score to 

*   higher number of nique words
*   higher relative amount of sentences within appropriate length

'docs_for_lime_BERT_&_LIME_2.csv' must be manually inserted into the content folder each session.

In [1]:
# Max: This code is unedited.

import pandas as pd # used in makind and editing the words data frame
import numpy as np 
#import json
#import csv


#import matplotlib.pyplot as plt # unnecessary
#import seaborn as sns # unnecessary

from tqdm import tqdm_notebook as tqdm # Progress Bar

from nltk.corpus import stopwords # removes words like “the”, “a”, “an”, and “in” # redundant?
from nltk.tokenize import word_tokenize # separates words into separate strings
from nltk.tokenize import sent_tokenize # separates sentences into separate strings
from nltk.stem import WordNetLemmatizer # identifies root word ("running" and "ran" are forms of "run")
import nltk # text reading
nltk.download('averaged_perceptron_tagger') # tags parts of speech (preposition, determiner, ...)

import spacy
#import math # unnecessary

#import string # unnecessary
#import sys # unnecessary
#import random # unnecessary

from collections import Counter # counts words, stores them in dictionary
from itertools import chain # Groups all the iterables together and produces a single iterable as output

#stop_words = set(stopwords.words('english')) # unnecessary
#lemmatizer = WordNetLemmatizer() # unnecessary

[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Unzipping taggers/averaged_perceptron_tagger.zip.


In [2]:
# Max: I added this code to Mishra's code.

#import nltk
nltk.download('stopwords') # Is this redundant?

from spacy.lang.en.examples import sentences

#nlp = spacy.load("en_core_web_sm") # nlp = spacy.load("en_trf_bertbaseuncased_lg")

# Code based off of:
# https://stackoverflow.com/questions/56927602/unable-to-load-the-spacy-model-en-core-web-lg-on-google-colab # Max: Should I delete this line?
import spacy.cli
spacy.cli.download("en_core_web_lg")
nlp = spacy.load("en_core_web_lg")

nltk.download('wordnet')

nltk.download('omw-1.4')

[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


✔ Download and installation successful
You can now load the package via spacy.load('en_core_web_lg')


[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


True

In [3]:
# Max: This code is unedited.

stop_words = set(stopwords.words('english')) # list of words like "a, but, an, ..."
lemmatizer = WordNetLemmatizer() # identifies word root

In [4]:
# Max: This code is edited and inserted.

df_sentences = pd.read_csv('docs_for_lime_BERT_&_LIME_2.csv')

In [5]:
# Max: This code is edited.

words=pd.DataFrame() # creates a data frame
frames=[df_sentences['documents']]
words['full'] = pd.concat(frames) # creates new column in the `words` dataframe
words.reset_index(inplace = True) # resets index
words['casewords'] = words['full'].str.lower() # creates new column, based off pre-extisting column
words['wsw'] = words['casewords'].apply(lambda x: ' '.join([word for word in x.split() if word not in (stop_words)])) # joins words into a string
words['nopunc'] = words['wsw'].str.replace("'", "") # new column, also removes punctiations
words['nopunc'] = words['nopunc'].str.replace(".", "") 
words['nopunc'] = words['nopunc'].str.replace(",", "")
w_tokenizer = nltk.tokenize.WhitespaceTokenizer() # extracts word or "tokens" and detaches whitespace
lemmatizer = nltk.stem.WordNetLemmatizer() # provides a synonym for an input word
def lemmatize_text(text): # finds roots of words in text
    return [lemmatizer.lemmatize(w) for w in w_tokenizer.tokenize(text)] 
words['text_lemmatized'] = words.nopunc.apply(lemmatize_text) # makes new column, with root word
words['final']="NaN" # "Not a Number", numeric data type, value is undefined or unrepresentable 
def final(x):
    makeitastring = ' '.join(map(str, x)) # applies str function to each x, each x separated by ' '
    makeitastring=makeitastring.replace(",","") # new column, also removes punctiations
    makeitastring=makeitastring.replace("'","")
    makeitastring=makeitastring.replace("[","")
    makeitastring=makeitastring.replace("]","")
    return makeitastring
for i in tqdm(range(len(words))): # range() creates a sequence. range(4) = [0,1,2,3], loading bar
    column=['text_lemmatized'] #
    row=[i]
    words.loc[i,'final']=final(words.loc[i,'text_lemmatized']) #.loc[] locates a group of rows and columns by a label
words['fullnopunc'] = words['full'].str.replace("'", "") # removes punctiations
words['fullnopunc'] = words['fullnopunc'].str.replace(".", "")
words['fullnopunc'] = words['fullnopunc'].str.replace(",", "")
words['label']='NaN'

<ipython-input-5-f6ff744740f6>:10: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  words['nopunc'] = words['nopunc'].str.replace(".", "")
<ipython-input-5-f6ff744740f6>:25: TqdmDeprecationWarning: This function will be removed in tqdm==5.0.0
Please use `tqdm.notebook.tqdm` instead of `tqdm.tqdm_notebook`
  for i in tqdm(range(len(words))): # range() creates a sequence. range(4) = [0,1,2,3], loading bar


  0%|          | 0/86 [00:00<?, ?it/s]

<ipython-input-5-f6ff744740f6>:30: FutureWarning: The default value of regex will change from True to False in a future version. In addition, single character regular expressions will *not* be treated as literal strings when regex=True.
  words['fullnopunc'] = words['fullnopunc'].str.replace(".", "")


In [ ]:
# `print(words)` shows what the words data frame looks like.

# print(words)

In [6]:
# Max: This code is unedited.

magwords=words.final.str.split(expand=True).stack().value_counts() # list words and its frequency
t1g=magwords.size/len(df_sentences)

In [7]:
# Max: This code is unedited.

g=[]
for i in range(len(words)):
    x=words.iloc[i]['fullnopunc']
    g.append(len(x.split()))
gstd=pd.Series(g).std() # standard deviation in word length

In [8]:
# Max: This code is unedited.

nrg=len(words)
drg=0
for i in range(len(g)): # sign function
    xg=g[i]
    xg=(xg-3)*(30-xg) # Mishra set 3 and 30 as his lower and upper legnth in word thresholds
    if(xg>0):
        drg=drg+1
    else:
        drg=drg-1

In [9]:
# Max: This code is unedited.

print(t1g) # T1
print(gstd) # T2
print(nrg)
print(drg)
t3g=drg/nrg
print(t3g) # T3

32.23255813953488
64.72230167252498
86
-84
-0.9767441860465116


In [10]:
dqic1g=t1g+(gstd*t3g)
print(dqic1g) # DQI c1 value

-30.984573726652307


Adapted DQI c1

In [11]:
print(t1g)

32.23255813953488


## How should the DQI c1 results be interpreted?

In his paper, Mishra provides a table to interpret DQI c1 and its components. He says that 7.6 is a "good" score and that 6.2 is a "bad" score, so should the -30.98 be interpreted?

Mishra further states that component T1 of DQI c1 is "good" above 1.9. The data yielded a value pf 32.2, meaning the data have a robust amount of unique words.

The bizarre -30.98 comes from the fact that Mishra tested his model on code where each obersation truly was only one sentence. For these data, each observation is actually composed of several sentences. This means that the standard variation of word length will vary greatly and that most sentences will far exceed Mishra's range of 3-30 words per sentence.

DQI c1 still provides valuable information as it shows that the data have a large number of unique words (which is one relevant dimension of robustness.